commit 54ddf2c8973663f04a820937001e248dda5b58e1
Author: Sage Weil <sweil@redhat.com>
Date:   Fri Nov 22 09:24:11 2019 -0600

    Revert "Merge pull request #16715 from adamemerson/wip-I-Object!"
    
    This reverts commit 669453138d89e0f797a1bd37f38a2d68e6aac366, reversing
    changes made to 36f5fcbb97eb2b1bceb526331eb3464f460fc701.
    
    Signed-off-by: Sage Weil <sage@redhat.com>
    
    - conflicts due to code rearrangement in 14b0db908f652032c358e419ffa90f5676698d0e

diff --git a/src/mgr/MgrStandby.cc b/src/mgr/MgrStandby.cc
index 913009e7e10..b021a182971 100644
--- a/src/mgr/MgrStandby.cc
+++ b/src/mgr/MgrStandby.cc
@@ -38,7 +38,7 @@
 
 MgrStandby::MgrStandby(int argc, const char **argv) :
   Dispatcher(g_ceph_context),
-  monc{g_ceph_context, poolctx},
+  monc{g_ceph_context},
   client_messenger(Messenger::create(
 		     g_ceph_context,
 		     cct->_conf.get_val<std::string>("ms_type"),
@@ -46,7 +46,7 @@ MgrStandby::MgrStandby(int argc, const char **argv) :
 		     "mgr",
 		     getpid(),
 		     0)),
-  objecter{g_ceph_context, client_messenger.get(), &monc, poolctx, 0, 0},
+  objecter{g_ceph_context, client_messenger.get(), &monc, NULL, 0, 0},
   client{client_messenger.get(), &monc, &objecter},
   mgrc(g_ceph_context, client_messenger.get(), &monc.monmap),
   log_client(g_ceph_context, client_messenger.get(), &monc.monmap, LogClient::NO_FLAGS),
@@ -115,8 +115,6 @@ int MgrStandby::init()
   client_messenger->add_dispatcher_tail(&client);
   client_messenger->start();
 
-  poolctx.start(2);
-
   // Initialize MonClient
   if (monc.build_initial_monmap() < 0) {
     client_messenger->shutdown();
@@ -171,7 +169,6 @@ int MgrStandby::init()
   // this method.
   monc.set_passthrough_monmap();
 
-  poolctx.start(1);
   client_t whoami = monc.get_global_id();
   client_messenger->set_myname(entity_name_t::MGR(whoami.v));
   monc.set_log_client(&log_client);
@@ -282,31 +279,17 @@ void MgrStandby::shutdown()
     }
 
     py_module_registry.shutdown();
-    // stop sending beacon first, i use monc to talk with monitors
-    timer.shutdown();
-    // client uses monc and objecter
-    client.shutdown();
-    mgrc.shutdown();
-    // Stop asio threads, so leftover events won't call into shut down
-    // monclient/objecter.
-    poolctx.finish();
-    // stop monc, so mon won't be able to instruct me to shutdown/activate after
-    // the active_mgr is stopped
-    monc.shutdown();
-    if (active_mgr) {
-      active_mgr->shutdown();
-    }
+
     // objecter is used by monc and active_mgr
     objecter.shutdown();
     // client_messenger is used by all of them, so stop it in the end
     client_messenger->shutdown();
   }));
 
-  // objecter is used by monc and active_mgr
-  objecter.shutdown();
-  // client_messenger is used by all of them, so stop it in the end
-  client_messenger->shutdown();
-  poolctx.finish();
+  // Then stop the finisher to ensure its enqueued contexts aren't going
+  // to touch references to the things we're about to tear down
+  finisher.wait_for_empty();
+  finisher.stop();
 }
 
 void MgrStandby::respawn()
diff --git a/src/mgr/MgrStandby.h b/src/mgr/MgrStandby.h
index cac31a57640..d79fef2d5e5 100644
--- a/src/mgr/MgrStandby.h
+++ b/src/mgr/MgrStandby.h
@@ -16,7 +16,6 @@
 #define MGR_STANDBY_H_
 
 #include "auth/Auth.h"
-#include "common/async/context_pool.h"
 #include "common/Finisher.h"
 #include "common/Timer.h"
 #include "common/LogClient.h"
@@ -40,7 +39,6 @@ public:
 			  const std::set <std::string> &changed) override;
 
 protected:
-  ceph::async::io_context_pool poolctx;
   MonClient monc;
   std::unique_ptr<Messenger> client_messenger;
   Objecter objecter;
